src/hotspot/share/runtime/lightweightSynchronizer.cpp

```diff
@@ -0,0 +1,1228 @@
+    ConcurrentTable::BulkDeleteTask clean_task(_table);
+    auto is_dead = [&](ObjectMonitor** monitor) {
+      return (*monitor)->object_is_dead();
+    };
+    auto do_nothing = [&](ObjectMonitor** monitor) {};
+    NativeHeapTrimmer::SuspendMark sm("ObjectMonitorTable");
+    return run_task(current, clean_task, "Clean", is_dead, do_nothing);
+  }
+
+  static bool resize(JavaThread* current) {
+    LogTarget(Info, monitortable) lt;
+    bool success = false;
+
+    if (should_grow()) {
+      lt.print("Start growing with load factor %f", get_load_factor());
+      success = grow(current);
+    } else {
+      if (!_table->is_max_size_reached() && Atomic::load(&_resize)) {
+        lt.print("WARNING: Getting resize hints with load factor %f", get_load_factor());
+      }
+      lt.print("Start cleaning with load factor %f", get_load_factor());
+      success = clean(current);
+    }
+
+    Atomic::store(&_resize, false);
+
+    return success;
+  }
+
+  static ObjectMonitor* monitor_put_get(Thread* current, ObjectMonitor* monitor, oop obj) {
+    // Enter the monitor into the concurrent hashtable.
+    ObjectMonitor* result = monitor;
+    Lookup lookup_f(obj);
+    auto found_f = [&](ObjectMonitor** found) {
+      assert((*found)->object_peek() == obj, "must be");
+      result = *found;
+    };
+    bool grow;
+    _table->insert_get(current, lookup_f, monitor, found_f, &grow);
+    verify_monitor_get_result(obj, result);
+    if (grow) {
+      try_notify_grow();
+    }
+    return result;
+  }
+
+  static bool remove_monitor_entry(Thread* current, ObjectMonitor* monitor) {
+    LookupMonitor lookup_f(monitor);
+    return _table->remove(current, lookup_f);
+  }
+
+  static bool contains_monitor(Thread* current, ObjectMonitor* monitor) {
+    LookupMonitor lookup_f(monitor);
+    bool result = false;
+    auto found_f = [&](ObjectMonitor** found) {
+      result = true;
+    };
+    _table->get(current, lookup_f, found_f);
+    return result;
+  }
+
+  static void print_on(outputStream* st) {
+    auto printer = [&] (ObjectMonitor** entry) {
+       ObjectMonitor* om = *entry;
+       oop obj = om->object_peek();
+       st->print("monitor " PTR_FORMAT " ", p2i(om));
+       st->print("object " PTR_FORMAT, p2i(obj));
+       assert(obj->mark().hash() == om->hash(), "hash must match");
+       st->cr();
+       return true;
+    };
+    if (SafepointSynchronize::is_at_safepoint()) {
+      _table->do_safepoint_scan(printer);
+    } else {
+      _table->do_scan(Thread::current(), printer);
+    }
+  }
+};
+
+ObjectMonitorTable::ConcurrentTable* ObjectMonitorTable::_table = nullptr;
+volatile size_t ObjectMonitorTable::_items_count = 0;
+size_t ObjectMonitorTable::_table_size = 0;
+volatile bool ObjectMonitorTable::_resize = false;
+
+ObjectMonitor* LightweightSynchronizer::get_or_insert_monitor_from_table(oop object, JavaThread* current, bool* inserted) {
+  assert(LockingMode == LM_LIGHTWEIGHT, "must be");
```

Comment:
_Must have missed pressing Comment_

The assert does not seem necessary, this member function is local to LightweightSynchronizer (private) and it will crash hard if it is called with a bad pointer for the out parameter.

It could have been a reference type here instead. I am not sure what is the best when it comes to out parameters, all styles have pros and cons. We seem to use all different combinations throughout hotspot.